Add useful diagnostic when trying to write through key path on private(set) var

sl opened the pull request:
This changes resolveImmutableBase so that instead of returning std::pair<Expr *, ValueDecl *>, it now returns std::pair<Expr*, Optional<OverloadChoice>>. It then uses this information to produce a better diagnostic.

Resolves [SR-10202](https://bugs.swift.org/browse/SR-10202).

sl committed:
Sema: Add a diagnostic for assigning using a subscript with a read-only keypath

sl committed:
Test: Update swift 5 keypath tests to reflect and test new behavior

sl commented:
@swift-ci please smoke test

sl commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -981,100 +981,128 @@ bool AssignmentFailure::diagnoseAsError() {
+        std::string message = "the key path";
+        if (auto *SE = dyn_cast<SubscriptExpr>(immInfo.first)) {
+          if (auto *tupleExpr = dyn_cast<TupleExpr>(SE->getIndex())) {
+            if (auto *DRE = dyn_cast<DeclRefExpr>(tupleExpr->getElement(0))) {
```
These nested ifs for all these `dyn_cast`s are kinda ugly, but I'm not sure if there's a nicer way to write this. Any suggestions would be appreciated 🙂

theblixguy replied:
You can combine some of them together. For example:

```c++
if (choice.hasValue() && !choice->isDecl()) { ... }
```

You can also write the casts as:

```c++
auto *SE = dyn_cast<SubscriptExpr>(immInfo.first);
if (!SE) return;

auto *TE = dyn_cast<TupleExpr>(SE->getIndex());
if (!TE) return;

...
```

sl replied:
I don't think i can combine `choice.hasValue()` with `!choice->isDecl()`, cause the following if depends on it being a decl and is also inside the `if (choice.hasValue()) {...}`. Also, in this case, we sadly can't just return if the cast fails because we want to keep going but just without setting the message. I guess there's not a super clean way to write it in this case so I'll keep it as is for now.

theblixguy replied:
I was referring to line 984

theblixguy replied:
Oh, you're right. You can skip the `return`'s and use `dyn_cast_or_null` instead. If any of the casts fails, your final expression/decl will also be `null`.

```c++
auto *SE = dyn_cast_or_null<SubscriptExpr>(immInfo.first);
auto *TE = dyn_cast_or_null<TupleExpr>(SE ? SE->getIndex() : nullptr);
...
```

Not a big fan of this though so I guess it's better to stick with the ifs!

sl replied:
Yeah, I suppose so. I hate seeing that many levels of nesting, but guess there's not anything super nice we can do there.

xedin replied:
Another way of doing this would be to extract this into `Expr *getKeyPathArgument(SubscriptExpr *expr)` either static function or a closure here.

xedin commented:
@sl You forgot to add a test-case from SR-10202

theblixguy commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -981,100 +981,128 @@ bool AssignmentFailure::diagnoseAsError() {
+    // Otherwise, we cannot resolve this because the available setter candidates
+    // are all mutating and the base must be mutating.  If we dug out a
+    // problematic decl, we can produce a nice tailored diagnostic.
+    if (auto *VD = dyn_cast_or_null<VarDecl>(choice->getDecl())) {
```
Another way would be:

```c++
if (!choice->isDecl()) return;
auto VD = dyn_cast<VarDecl>(choice->getDecl);
```

sl replied:
Sadly I don't think we can do this either.  If the choice isn't a decl, we don't wanna return, we just want to skip over the next section. Af this, we try to diagnose the issue using information about the expression so that's what we're trying to skip to. If we have to add the level of indentation anyway, we might as well combine the cast and check in one line.

theblixguy replied:
Whoops, I misread the code, sorry. Seems like my suggestions are useless because we _don't_ want to return early; we have to emit a different diagnostic.

xedin replied:
Can use `dyn_cast<VarDecl>` here because you have already checked that choice has a declaration.

sl replied:
@theblixguy Yeah, that sounds like it wouldn't be worth the effort given that the change to the readability here would only be marginal. Good idea though 🙂

sl replied:
Changed that to use dyn_cast, so this should be all set 👍

theblixguy commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -981,100 +981,128 @@ bool AssignmentFailure::diagnoseAsError() {
+      auto type = getType(immInfo.first);
+
+      if (isKnownKeyPathType(type))
+        message += " is a read-only key path";
```
You could probably do this by adding `select` to the diagnostic message, although it might be overkill if you have so many possibilities here.

xedin replied:
That select unfortunately is going to be pretty big, I'd suggest to drop trailing "key path" in this case though.

theblixguy commented on test/Constraints/keypath_swift_5.swift:
```diff
@@ -19,3 +19,18 @@ func test() {
+  let path = \T.a
+  var t = T(a: 3)
+  t[keyPath: path] = 4 // expected-error {{cannot assign through subscript: 'path' is a read-only key path}}
+  t[keyPath: \T.a] = 4 // expected-error {{cannot assign through subscript: the key path is a read-only key path}}
```
`the key path is a read-only key path` reads a bit weird to me - maybe this could be `'T.a' is a read-only key-path` and then it would match the diagnostic above.

sl replied:
The current behavior for similar errors is to give the error `"cannot assign through subscript: 'x' is a read-only key path"` if there's a declaration that we can get the name from, and give "cannot assign through subscript: immutable key path". As such, adding additional information to the error message in the second case would be an improvement aside from the bug fix. I'd be happy to change the wording as I agree it is a bit clunky (maybe just "cannot assign through subscript: is read-only" might be better?) but I feel like adding the keyPath name here might be outside the scope of this PR. Thoughts @xedin?

xedin replied:
Yeah, I think it would make sense to just drop "key path" at the end, so it would say `key path is read-only`.

sl commented:
@xedin Added tests for the new behavior 👍

sl committed:
Test / Sema: Change the wording of assignment through read-only key path error

sl committed:
Sema: Change cast to use dyn_cast instead of dyn_cast_or_null

sl committed:
Sema: Move getting key-path argument logic to a closure

sl commented:
@swift-ci please smoke test

sl committed:
Test: Update swift 4 keypath tests to reflect changed error message

sl commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -1227,40 +1259,63 @@ void AssignmentFailure::fixItChangeInoutArgType(const Expr *arg,
   // Provide specific diagnostics for assignment to subscripts whose base expr
   // is known to be an rvalue.
   if (auto *SE = dyn_cast<SubscriptExpr>(expr)) {
+
```
Nit - empty line

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -1227,40 +1259,63 @@ void AssignmentFailure::fixItChangeInoutArgType(const Expr *arg,
   auto *DC = getDC();
   expr = expr->getValueProvidingExpr();
 
+  auto isImmutable = [&](ValueDecl *decl) {
```
I don't think we need to capture anything besides DC here.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -1227,40 +1259,63 @@ void AssignmentFailure::fixItChangeInoutArgType(const Expr *arg,
+    Optional<OverloadChoice> member;
+
+    auto loc = cs.getConstraintLocator(SE, ConstraintLocator::SubscriptMember);
+    member = getMemberRef(loc);
```
Can inline this locator call into `getMemberRef` since it's not going to be used anywhere else.

xedin commented on test/Constraints/keypath_swift_5.swift:
```diff
@@ -7,15 +7,30 @@ struct S {
 
     S()[keyPath: \.i] = 1
-    // expected-error@-1 {{cannot assign through subscript: immutable key path}}
+    // expected-error@-1 {{cannot assign through subscript: is read-only}}
```
Hm, IMHO this needs `key path` bit in this diagnostic.

xedin commented on test/Constraints/keypath_swift_5.swift:
```diff
@@ -7,15 +7,30 @@ struct S {
 
   C()[keyPath: \.i] = 1
-  // expected-error@-1 {{cannot assign through subscript: immutable key path}}
+  // expected-error@-1 {{cannot assign through subscript: is read-only}}
```
And this one too.

xedin commented on test/Constraints/keypath_swift_5.swift:
```diff
@@ -7,15 +7,30 @@ struct S {
+
+
+struct T {
+    private(set) var a: Int
```
Nit: two space indentation in tests.

xedin commented on test/Constraints/keypath_swift_5.swift:
```diff
@@ -7,15 +7,30 @@ struct S {
+  let path = \T.a
+  var t = T(a: 3)
+  t[keyPath: path] = 4 // expected-error {{cannot assign through subscript: 'path' is read-only}}
+  t[keyPath: \T.a] = 4 // expected-error {{cannot assign through subscript: is read-only}}
```
"key path is read-only"

xedin commented on test/expr/unary/keypath/keypath.swift:
```diff
@@ -401,8 +401,8 @@ func testKeyPathSubscriptMetatype(readonly: Z.Type, writable: inout Z.Type,
 
-  readonly[keyPath: kp] = sink // expected-error{{cannot assign through subscript: 'kp' is a read-only key path}}
-  writable[keyPath: kp] = sink // expected-error{{cannot assign through subscript: 'kp' is a read-only key path}}
+  readonly[keyPath: kp] = sink // expected-error{{cannot assign through subscript: 'kp' is read-only}}
```
I like the old wording better here too because otherwise it's unclear what is `kp`. When I said - let's drop "key path" at the end - I meant only in the situation when we'd already have that in the beginning of the message, one of the messages was something like - `key path is read-only key path` so `key path` bit doesn't really make much sense to repeat.

sl replied:
Ahh, I misunderstood your comment. I’ll switch that around

xedin reviewed (changes requested):
Overall looks great! Some messages still need more work and there are a couple nits in the code.

sl committed:
Sema / Test: Revert the error for assignment through read-only key path

This updates the error message so that in the case where we can find a
Decl, it gives the error "cannot assign through subscript: 'name' is a
read-only key path", and in the case where there's no associated Decl, gives the
error message "cannot assign through subscript: key path is read-only".

Additionally updates tests with the new error messages and formats all changes.

sl commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -981,101 +981,133 @@ bool AssignmentFailure::diagnoseAsError() {
+
+    auto getKeyPathArgument = [](SubscriptExpr *expr) {
+      auto *TE = dyn_cast<TupleExpr>(expr->getIndex());
+      assert(TE->getNumElements() == 1);
```
Nit: Would be great to assert that element has a label `keyPath:` just like original code in `resolveImmutableBase` did.

xedin reviewed (approved):
LGTM! Please fix this small thing I mentioned and feel free to merge.

sl committed:
Sema: Add additional assertion before providing keypath application diagnostic

sl commented:
@swift-ci please smoke test

theblixguy commented:
Thanks @sl for the new diagnostic! 👍
